### ISSUE TITLE
gnupg: Gnome3 pinentry fails

### ISSUE THREAD

**Issue by NinjaTrappeur** [NixOS/nixpkgs#112914]
**Describe the bug**

I just bumped Nixpkgs (d8d61114a6abfe54fcacfdd9631bd3869a319960 -> 8c8731330b53ba0061686f36f10f101e662a4717), this bump broke my gnupg Gnome3 pinentry prompt. It now fails with:

```
No Gcr System Prompter available, falling back to curses
```

It then fallbacks to the ncurse prompt, which is sadly not an option for emacs users :(

I am using Sway wm. @Mic92 is experiencing the same bug.

My current workaround is to rather use the qt pinentry prompt through

```
  programs.gnupg.agent = {
    enable = true;
    pinentryFlavor = "qt";
  };
```

**Additional context**

DBus message dump using `busctl monitor --user` (the `$REDACTED` parts have been redacted):

```
‣ Type=method_call  Endian=l  Flags=0  Version=1 Cookie=1
  Sender=:1.37  Destination=org.freedesktop.DBus  Path=/org/freedesktop/DBus  Interface=org.freedesktop.DBus  Member=Hello
  UniqueName=:1.37
  MESSAGE "" {
  };

‣ Type=method_return  Endian=l  Flags=1  Version=1 Cookie=1  ReplyCookie=1
  Sender=org.freedesktop.DBus  Destination=:1.37
  MESSAGE "s" {
          STRING ":1.37";
  };

‣ Type=signal  Endian=l  Flags=1  Version=1 Cookie=57
  Sender=org.freedesktop.DBus  Path=/org/freedesktop/DBus  Interface=org.freedesktop.DBus  Member=NameOwnerChanged
  MESSAGE "sss" {
          STRING ":1.37";
          STRING "";
          STRING ":1.37";
  };

‣ Type=signal  Endian=l  Flags=1  Version=1 Cookie=2
  Sender=org.freedesktop.DBus  Destination=:1.37  Path=/org/freedesktop/DBus  Interface=org.freedesktop.DBus  Member=NameAcquired
  MESSAGE "s" {
          STRING ":1.37";
  };

‣ Type=method_call  Endian=l  Flags=0  Version=1 Cookie=2
  Sender=:1.37  Destination=org.freedesktop.DBus  Path=/org/freedesktop/DBus  Interface=org.freedesktop.DBus  Member=AddMatch
  UniqueName=:1.37
  MESSAGE "s" {
          STRING "type='signal',sender='org.freedesktop.DBus',interface='org.freedesktop.DBus',member='NameOwnerChanged',path='/org/freedesktop/DBus',arg0='org.gnome.keyring.SystemPrompter'";
  };

‣ Type=method_return  Endian=l  Flags=1  Version=1 Cookie=3  ReplyCookie=2
  Sender=org.freedesktop.DBus  Destination=:1.37
  MESSAGE "" {
  };

‣ Type=method_call  Endian=l  Flags=0  Version=1 Cookie=3
  Sender=:1.37  Destination=org.freedesktop.DBus  Path=/org/freedesktop/DBus  Interface=org.freedesktop.DBus  Member=GetNameOwner
  UniqueName=:1.37
  MESSAGE "s" {
          STRING "org.gnome.keyring.SystemPrompter";
  };

‣ Type=error  Endian=l  Flags=1  Version=1 Cookie=4  ReplyCookie=3
  Sender=org.freedesktop.DBus  Destination=:1.37
  ErrorName=org.freedesktop.DBus.Error.NameHasNoOwner  ErrorMessage="Could not get owner of name 'org.gnome.keyring.SystemPrompter': no such name"
  MESSAGE "s" {
          STRING "Could not get owner of name 'org.gnome.keyring.SystemPrompter': no such name";
  };

‣ Type=method_call  Endian=l  Flags=0  Version=1 Cookie=4
  Sender=:1.37  Destination=org.gnome.keyring.SystemPrompter  Path=/org/gnome/keyring/Prompter  Interface=org.gnome.keyring.internal.Prompter  Member=BeginPrompting
  UniqueName=:1.37
  MESSAGE "o" {
          OBJECT_PATH "/org/gnome/keyring/Prompt/p0";
  };

‣ Type=error  Endian=l  Flags=1  Version=1 Cookie=5  ReplyCookie=4
  Sender=org.freedesktop.DBus  Destination=:1.37
  ErrorName=org.freedesktop.DBus.Error.Spawn.ChildExited  ErrorMessage="Process org.gnome.keyring.SystemPrompter exited with status 1"
  MESSAGE "s" {
          STRING "Process org.gnome.keyring.SystemPrompter exited with status 1";
  };

‣ Type=signal  Endian=l  Flags=1  Version=1 Cookie=5
  Sender=org.freedesktop.DBus  Destination=:1.37  Path=/org/freedesktop/DBus  Interface=org.freedesktop.DBus  Member=NameLost
  MESSAGE "s" {
          STRING ":1.37";
  };

‣ Type=signal  Endian=l  Flags=1  Version=1 Cookie=58
  Sender=org.freedesktop.DBus  Path=/org/freedesktop/DBus  Interface=org.freedesktop.DBus  Member=NameOwnerChanged
  MESSAGE "sss" {
          STRING ":1.37";
          STRING ":1.37";
          STRING "";
  };

‣ Type=method_call  Endian=l  Flags=0  Version=1 Cookie=1
  Sender=:1.38  Destination=org.freedesktop.DBus  Path=/org/freedesktop/DBus  Interface=org.freedesktop.DBus  Member=Hello
  UniqueName=:1.38
  MESSAGE "" {
  };

‣ Type=method_return  Endian=l  Flags=1  Version=1 Cookie=1  ReplyCookie=1
  Sender=org.freedesktop.DBus  Destination=:1.38
  MESSAGE "s" {
          STRING ":1.38";
  };

‣ Type=signal  Endian=l  Flags=1  Version=1 Cookie=59
  Sender=org.freedesktop.DBus  Path=/org/freedesktop/DBus  Interface=org.freedesktop.DBus  Member=NameOwnerChanged
  MESSAGE "sss" {
          STRING ":1.38";
          STRING "";
          STRING ":1.38";
  };

‣ Type=signal  Endian=l  Flags=1  Version=1 Cookie=2
  Sender=org.freedesktop.DBus  Destination=:1.38  Path=/org/freedesktop/DBus  Interface=org.freedesktop.DBus  Member=NameAcquired
  MESSAGE "s" {
          STRING ":1.38";
  };

‣ Type=method_call  Endian=l  Flags=0  Version=1 Cookie=2
  Sender=:1.38  Destination=org.freedesktop.DBus  Path=/org/freedesktop/DBus  Interface=org.freedesktop.DBus  Member=AddMatch
  UniqueName=:1.38
  MESSAGE "s" {
          STRING "type='signal',sender='org.freedesktop.secrets',interface='org.freedesktop.DBus.Properties',member='PropertiesChanged',path='/org/freedesktop/secrets',arg0='org.freedesktop.Secret.Service'";
  };

‣ Type=method_return  Endian=l  Flags=1  Version=1 Cookie=3  ReplyCookie=2
  Sender=org.freedesktop.DBus  Destination=:1.38
  MESSAGE "" {
  };

‣ Type=method_call  Endian=l  Flags=0  Version=1 Cookie=3
  Sender=:1.38  Destination=org.freedesktop.DBus  Path=/org/freedesktop/DBus  Interface=org.freedesktop.DBus  Member=AddMatch
  UniqueName=:1.38
  MESSAGE "s" {
          STRING "type='signal',sender='org.freedesktop.secrets',interface='org.freedesktop.Secret.Service',path='/org/freedesktop/secrets'";
  };

‣ Type=method_return  Endian=l  Flags=1  Version=1 Cookie=4  ReplyCookie=3
  Sender=org.freedesktop.DBus  Destination=:1.38
  MESSAGE "" {
  };

‣ Type=method_call  Endian=l  Flags=0  Version=1 Cookie=4
  Sender=:1.38  Destination=org.freedesktop.DBus  Path=/org/freedesktop/DBus  Interface=org.freedesktop.DBus  Member=AddMatch
  UniqueName=:1.38
  MESSAGE "s" {
          STRING "type='signal',sender='org.freedesktop.DBus',interface='org.freedesktop.DBus',member='NameOwnerChanged',path='/org/freedesktop/DBus',arg0='org.freedesktop.secrets'";
  };

‣ Type=method_return  Endian=l  Flags=1  Version=1 Cookie=5  ReplyCookie=4
  Sender=org.freedesktop.DBus  Destination=:1.38
  MESSAGE "" {
  };

‣ Type=method_call  Endian=l  Flags=0  Version=1 Cookie=5
  Sender=:1.38  Destination=org.freedesktop.DBus  Path=/org/freedesktop/DBus  Interface=org.freedesktop.DBus  Member=StartServiceByName
  UniqueName=:1.38
  MESSAGE "su" {
          STRING "org.freedesktop.secrets";
          UINT32 0;
  };

‣ Type=method_return  Endian=l  Flags=1  Version=1 Cookie=6  ReplyCookie=5
  Sender=org.freedesktop.DBus  Destination=:1.38
  MESSAGE "u" {
          UINT32 2;
  };

‣ Type=method_call  Endian=l  Flags=0  Version=1 Cookie=6
  Sender=:1.38  Destination=org.freedesktop.DBus  Path=/org/freedesktop/DBus  Interface=org.freedesktop.DBus  Member=GetNameOwner
  UniqueName=:1.38
  MESSAGE "s" {
          STRING "org.freedesktop.secrets";
  };

‣ Type=method_return  Endian=l  Flags=1  Version=1 Cookie=7  ReplyCookie=6
  Sender=org.freedesktop.DBus  Destination=:1.38
  MESSAGE "s" {
          STRING ":1.10";
  };

‣ Type=method_call  Endian=l  Flags=0  Version=1 Cookie=7
  Sender=:1.38  Destination=:1.10  Path=/org/freedesktop/secrets  Interface=org.freedesktop.DBus.Properties  Member=GetAll
  UniqueName=:1.38
  MESSAGE "s" {
          STRING "org.freedesktop.Secret.Service";
  };

‣ Type=method_return  Endian=l  Flags=1  Version=1 Cookie=12  ReplyCookie=7
  Sender=:1.10  Destination=:1.38
  UniqueName=:1.10
  MESSAGE "a{sv}" {
          ARRAY "{sv}" {
                  DICT_ENTRY "sv" {
                          STRING "Collections";
                          VARIANT "ao" {
                                  ARRAY "o" {
                                          OBJECT_PATH "/org/freedesktop/secrets/collection/session";
                                          OBJECT_PATH "/org/freedesktop/secrets/collection/login";
                                  };
                          };
                  };
          };
  };

‣ Type=method_call  Endian=l  Flags=0  Version=1 Cookie=8
  Sender=:1.38  Destination=:1.10  Path=/org/freedesktop/secrets  Interface=org.freedesktop.Secret.Service  Member=OpenSession
  UniqueName=:1.38
  MESSAGE "sv" {
          STRING "dh-ietf1024-sha256-aes128-cbc-pkcs7";
          VARIANT "ay" {
                  ARRAY "y" {
                      $REDACTED
                  };
          };
  };

‣ Type=method_return  Endian=l  Flags=1  Version=1 Cookie=13  ReplyCookie=8
  Sender=:1.10  Destination=:1.38
  UniqueName=:1.10
  MESSAGE "vo" {
          VARIANT "ay" {
                  ARRAY "y" {
                     $REDACTED
                  };
          };
          OBJECT_PATH "/org/freedesktop/secrets/session/s3";
  };

‣ Type=method_call  Endian=l  Flags=0  Version=1 Cookie=9
  Sender=:1.38  Destination=:1.10  Path=/org/freedesktop/secrets  Interface=org.freedesktop.Secret.Service  Member=SearchItems
  UniqueName=:1.38
  MESSAGE "a{ss}" {
          ARRAY "{ss}" {
                  DICT_ENTRY "ss" {
                          STRING "keygrip";
                          STRING "n/$REDACTED_MY_GPG_KEY";
                  };
                  DICT_ENTRY "ss" {
                          STRING "xdg:schema";
                          STRING "org.gnupg.Passphrase";
                  };
          };
  };

‣ Type=method_return  Endian=l  Flags=1  Version=1 Cookie=14  ReplyCookie=9
  Sender=:1.10  Destination=:1.38
  UniqueName=:1.10
  MESSAGE "aoao" {
          ARRAY "o" {
          };
          ARRAY "o" {
          };
  };

‣ Type=signal  Endian=l  Flags=1  Version=1 Cookie=6
  Sender=org.freedesktop.DBus  Destination=:1.38  Path=/org/freedesktop/DBus  Interface=org.freedesktop.DBus  Member=NameLost
  MESSAGE "s" {
          STRING ":1.38";
  };

‣ Type=signal  Endian=l  Flags=1  Version=1 Cookie=60
  Sender=org.freedesktop.DBus  Path=/org/freedesktop/DBus  Interface=org.freedesktop.DBus  Member=NameOwnerChanged
  MESSAGE "sss" {
          STRING ":1.38";
          STRING ":1.38";
          STRING "";
  };

```

I feel like the 

> STRING "Could not get owner of name 'org.gnome.keyring.SystemPrompter': no such name";

part might be relevant here.

-----

hexa- was pointing to this potentially recent patch on IRC: https://lists.gnupg.org/pipermail/gnupg-devel/2016-November/032058.html . I lack some domain knowledge to determine whether this is might be the root cause of this problem or not.

Previous chat on IRC: https://logs.nix.samueldr.com/nixos/2021-02-12#4618918;

**Notify maintainers**

@NixOS/gnome

**Comment by plabadens** [comment]
Having the same issue after switching from `nixos-20.09` to `nixos-unstable`.

**Comment by Ma27** [comment]
Same here, but I only updated to a more recent 20.09. Will try to find out what's going on here.

**Comment by Ma27** [comment]
Ok the issue on my side was that I was using `sway` from `nixpkgs-unstable` (i.e. sway 1.6 now). Going back to 1.5 fixed the problem for me.

**Comment by jtojnar** [comment]
Do you have `services.gnome.gnome-keyring.enable = true`?

**Comment by primeos** [comment]
I think/hope this issue can be closed now. Our new NixOS VM test now verifies that pinentry works: https://github.com/NixOS/nixpkgs/blob/85662edbfd9829a8182dc9641c2941d44d4a7488/nixos/tests/sway.nix#L86

The comments from Apr 10 should be fixed by #119445 but the original bug report is from Feb 12, i.e. before the update to Sway 1.6 (not sure what happened back then and if it's still an issue).

**Comment by plabadens** [comment]
Yes, I can confirm that `pinentry-gnome3` now works fine in combination with `gcr`.
